docs/guides/integration/alternative-indexes.md

```diff
@@ -168,6 +168,17 @@ export UV_PUBLISH_PASSWORD="$AWS_CODEARTIFACT_TOKEN"
 uv publish
 ```
 
+## GitLab
+
+If you want to add a dependency from a private GitLab package index, you need to force uv to pass credentials by adding the token name directly to the index's URL:
+
+```toml
+[[tool.uv.index]]
+url = "https://GITLAB_TOKEN@$gitlab.com/api/v4/{PROJECT-OR-GROUP}/-/packages/pypi/simple"
+```
+
+Since the token name is hardcoded in the pyproject.toml file, all users must set up a personal access token with the same name (GITLAB_TOKEN in the example) before installing dependencies.
+
```

Comment:
It might be preferred to suggest using [UV_INDEX_{name}_USERNAME](https://docs.astral.sh/uv/configuration/environment/#uv_index_name_username) and [UV_INDEX_{name}_USERNAME](https://docs.astral.sh/uv/configuration/environment/#uv_index_name_username) for the credentialed pieces.

Reply:
Yeah, that too. With the credentials and without token name in URL, uv will first try unauthroized request, receive 404 and call uncle. I'll reword it for better clarity. Any suggestions?

Reply:
https://github.com/astral-sh/uv/blob/81569c47bfa91b24ff0712baf1c001ef9604676e/crates/uv-auth/src/middleware.rs#L148-L160

Are you sure the 404 is the problem?  It is explicitly one of the status codes that UV will retry with auth for.

Reply:
What I'm sure is that adding token name fixes the issue described in the linked report. Take a look at the logs I've linked there. @zanieb seems to know the underlying cause.

Reply:
For what it is worth I ran into a very similar issue to the one that you had in terms of uv output, and determined that a strangely configured python package index was returning 200 with no auth, and not listing any versions unless auth was provided.  I may try to implement one of the configuration options discussed in the issue and raise a PR, but I am hoping that I can get the creator of the index I need to use at work to simply fix the probable permissions issue.

Reply:
Adding the token name fixes it because we skip trying an unauthenticated request.

My earlier comments were incorrect regarding the 404, thanks @Raymi306 for pointing that out — we will search for credentails on a 404. It looks like GitLab is actually returning an okay response? The relevant logs being:

```
TRACE No credentials in cache for URL https://gitlab.com/api/v4/groups/${ID}/-/packages/pypi/simple/livity-airtable/
TRACE Attempting unauthenticated request for https://gitlab.com/api/v4/groups/${ID}/-/packages/pypi/simple/livity-airtable/
TRACE cached request https://gitlab.com/api/v4/groups/${ID}/-/packages/pypi/simple/livity-airtable/ is storable because its response has a 'public' cache-control directive
TRACE Received package metadata for: livity-airtable
TRACE Selecting candidate for livity-airtable with range * with 0 remote versions
```

Sorry for misleading you there, I'm not sure why I thought we only retried on 401..

Reply:
Cool, thanks for clearing it out.

> strangely configured python package index was returning 200

Hmm, the free gitlab.com offering has only one config option in the package registry - forwarding requests to PyPI. It actually might be the issue. I'll investigate it further.

Reply:
Thanks @rafalkrupinski!